lua/lspconfig/manager.lua

```diff
@@ -150,88 +128,46 @@ function M:_start_new_client(bufnr, new_config, root_dir, single_file)
     new_config.workspace_folders = nil
   end
 
-  -- TODO: Replace lsp.start_client with lsp.start
-  local client_id, err = lsp.start_client(new_config)
-  if not client_id then
-    if err then
-      vim.notify(err, vim.log.levels.WARN)
-    end
-    return
-  end
-  self:_attach_and_cache(bufnr, root_dir, client_id)
-end
+  local client_id = lsp.start(new_config, {
+    bufnr = bufnr,
+    silent = silent,
+    reuse_client = function(existing_client)
+      if (self._clients[root_dir] or {})[existing_client.name] then
+        self:_notify_workspace_folder_added(root_dir, existing_client)
+        return true
+      end
 
---- @private
---- @param bufnr integer
---- @param new_config lspconfig.Config
---- @param root_dir string
---- @param client vim.lsp.Client
---- @param single_file boolean
-function M:_attach_or_spawn(bufnr, new_config, root_dir, client, single_file)
-  if check_in_workspace(client, root_dir) then
-    return self:_attach_and_cache(bufnr, root_dir, client.id)
-  end
+      for _, dir_clients in pairs(self._clients) do
+        if dir_clients[existing_client.name] then
+          self:_notify_workspace_folder_added(root_dir, existing_client)
+          return true
+        end
+      end
 
-  local supported = vim.tbl_get(client, 'server_capabilities', 'workspace', 'workspaceFolders', 'supported')
-  if supported then
-    return self:_register_workspace_folders(bufnr, root_dir, client)
+      return false
+    end,
+  })
+  if client_id then
+    self:_cache_client(root_dir, assert(lsp.get_client_by_id(client_id)))
   end
-  self:_start_new_client(bufnr, new_config, root_dir, single_file)
-end
-
---- @private
---- @param bufnr integer
---- @param new_config lspconfig.Config
---- @param root_dir string
---- @param client vim.lsp.Client
---- @param single_file boolean
-function M:_attach_after_client_initialized(bufnr, new_config, root_dir, client, single_file)
```

Comment:
When neovim reads the session it may contain multiple projects. This needs to be checked after the first server start.

Reply:
The logic of `_attach_or_spawn` is:
1. Call `lsp.buf_attach_client` (which neovim does inside `vim.lsp.start` so no need to do it here)
2. Add the client to the `_clients` cache
3. Send a `workspace/didChangeWorkspaceFolders` notification if the server supports it. 

The implementation in this PR keeps 2 and 3, event after the server is first initialized. Or are you referring to something else?

Reply:
vsplit has two buffers in two projects. mksession starts neovim with -S session.vim. The first buffer triggers FileType to start a server. Note that this is an asynchronous server that needs initialization time. At this point you are not sure if the server supports workspacefolder. But the second buffer has triggered FileType again. The embed process is processed continuously. So the second buffer should wait for the server started by the first buffer to complete initialization and check whether to connect to it or start a new one.

Reply:
I see. I've replaced the code using the timer by adding [this hook](https://github.com/neovim/nvim-lspconfig/blob/2eccb418f2f972ad1167f8491a8180acdeb02384/lua/lspconfig/manager.lua#L105) to `on_init`.

Reply:
if I remember correctly, this is a hook for modifying the parameters of the server itself. It runs after the root judgment. The problem is that if the server has not initialized the second buffer, the same event loop will run the luanch function. Can we know that the first server started supports workspacefolder? Shouldn't there be a synchronous waiting problem here? I have forgotten some of the code here. I simply tested this Pr code and it works as I expected. But I still don't understand the logic here.